metrics/scope.go

```diff
@@ -1,117 +1,87 @@
-
 package metrics
 
 import (
 	"strings"
 	"time"
 
-	"github.com/cactus/go-statsd-client/statsd"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
 // Scope is a stats collector that will prefix the name the stats it
 // collects.
 type Scope interface {
 	NewScope(scopes ...string) Scope
-	Scope() string
 
 	Inc(stat string, value int64) error
-	Dec(stat string, value int64) error
 	Gauge(stat string, value int64) error
 	GaugeDelta(stat string, value int64) error
 	Timing(stat string, delta int64) error
 	TimingDuration(stat string, delta time.Duration) error
-	Set(stat string, value string) error
 	SetInt(stat string, value int64) error
-	Raw(stat string, value string) error
 }
 
-// StatsdScope is a Scope that sends data to statsd with a prefix added to the
-// stat names.
-type StatsdScope struct {
-	prefix  string
-	statter statsd.Statter
+// promScope is a Scope that sends data to Prometheus
+type promScope struct {
+	*autoRegisterer
+	prefix     string
+	registerer prometheus.Registerer
 }
 
-var _ Scope = &StatsdScope{}
+var _ Scope = &promScope{}
 
-// NewStatsdScope returns a StatsdScope that prefixes stats it collections with
-// the scopes given joined together by periods
-func NewStatsdScope(statter statsd.Statter, scopes ...string) *StatsdScope {
-	return &StatsdScope{
-		prefix:  strings.Join(scopes, ".") + ".",
-		statter: statter,
+// NewPromScope returns a Scope that sends data to Prometheus
+func NewPromScope(registerer prometheus.Registerer, scopes ...string) Scope {
+	return &promScope{
+		prefix:         strings.Join(scopes, ".") + ".",
+		autoRegisterer: newAutoRegisterer(registerer),
 	}
 }
 
 // NewNoopScope returns a Scope that won't collect anything
 func NewNoopScope() Scope {
-	c, _ := statsd.NewNoopClient()
-	return NewStatsdScope(c, "noop")
+	return NewPromScope(prometheus.NewRegistry())
 }
 
 // NewScope generates a new Scope prefixed by this Scope's prefix plus the
 // prefixes given joined by periods
-func (s *StatsdScope) NewScope(scopes ...string) Scope {
+func (s *promScope) NewScope(scopes ...string) Scope {
 	scope := strings.Join(scopes, ".")
-	return NewStatsdScope(s.statter, s.prefix+scope)
-}
-
-// Scope returns the current string prefix (except for the final period) that
-// stats will receive
-func (s *StatsdScope) Scope() string {
-	return s.prefix[:len(s.prefix)-1]
+	return NewPromScope(s.registerer, s.prefix+scope)
 }
 
 // Inc increments the given stat and adds the Scope's prefix to the name
-func (s *StatsdScope) Inc(stat string, value int64) error {
-	autoCounter(s.prefix + stat).Add(float64(1))
-	return s.statter.Inc(s.prefix+stat, value, 1.0)
-}
-
-// Dec decrements the given stat and adds the Scope's prefix to the name
-func (s *StatsdScope) Dec(stat string, value int64) error {
-	return s.statter.Dec(s.prefix+stat, value, 1.0)
+func (s *promScope) Inc(stat string, value int64) error {
```

Comment:
Should value be replaced with a `_` since it isn't used for the .`Add()` call on the autoCounter?